index.js

```diff
@@ -109,7 +150,22 @@ Plugin.prototype.toString = function() {
 // This indirection allows subclasses like broccoli-caching-writer to hook
 // into calls from the builder, by returning { build: someFunction }
 Plugin.prototype.getCallbackObject = function() {
-  return this
+  var plugin = this;
+
+  if (plugin._sideEffectFree === false) {
+    return this;
+  }
+
+  return {
+    build: function() {
+      if (plugin._inputsHaveRevisions() || isNaN(plugin._revision)) {
+        if (isNaN(plugin._revision)) {
```

Comment:
Can we pull this out and move it to the root of this function (out of the nested `if` position)?